design-docs/modes.md

```diff
@@ -0,0 +1,638 @@
+
+- Modes are usually registered via annotation of classes.  These annotations may specify a name (or default to the class name), group name (for grouping of routines in the selection list), and description.  Annotations are used to specify whether the class is a teleop, auto, or test mode.
+
+- Modes may also be registered via annotation of functions (in the Robot class only) or via explicit function calls.  As C++ does not support annotations, function call registration is the only available method in that language.
+
+- For maximum flexibility, all code in the robot project has access to the enable/disable state, the overall robot teleop/auto/test mode, and the selected modes for teleop, auto, and test (even when the robot is disabled).
+
+- Modes may be periodic or linear (or custom).  The Robot base class handles switching between modes and mode object instance creation.  Mode objects are constructed when the drop-down selection is made in the DS and run when the robot is enabled.
+
+- The command-based framework provides convenience functions to link triggers to specific modes.  These triggers can then be linked to start specific commands (for auto modes).  To support multiple teleop modes, joystick bindings can be configured to only be active in specific modes.
+
+## Driver Station
+
+For reference, the FRC driver station is shown below.  This illustrates the enable/disable control (2) as well as the teleop/auto/practice/test selector (1).  For this design, "practice" is renamed to "match" for clarity.
+
+![Driver Station GUI](https://docs.wpilib.org/en/stable/_images/ds-operation-tab.webp)
+
+One or two drop-down selectors are added to this main screen.  One selector is displayed in teleop, auto, and test modes; two selectors (one for auto and one for teleop) is displayed in match mode (either manually selected by clicking "match," or when the DS is FMS-attached).
+
+The drop-down lists have the modes (as defined by user code) organized into groups, similar to the below.  The groups and modes are sorted.
+
+![Combo box with groups](https://i.sstatic.net/4Cmfk.gif)
+
+Only the routines appropriate to the mode are shown in the corresponding list.  For example, if the "Test" option is selected, a single drop-down list with only test modes listed will be shown.
+
+## Robot Code User-Facing API
+
+Java is used for illustrative purposes.  Python and C++ should follow a similar structure, except that C++ doesn't support annotations.
+
+### RobotBase
+
+The `RobotBase` class is the base class for the user's `Robot` class.  It also implements the private library machinery for robot startup and robot execution (including creating and transitioning between modes in accordance with the mode lifecycle, as described in the following section).
+
```

Comment:
This is not what I was thinking at all.   I was thinking of each "OpMode" (or whatever it gets named) having a class member that got a singleton instance of the "Robot" class.

Reply:
That's probably the biggest difference from current FTC opmodes.  The Robot instance is constructed first (by main()) and modes are constructed later.  Robot is effectively a singleton; it's a single object that persists through all modes that is provided to each mode's constructor (and the mode constructor stores it into a member variable).